### PR TITLE
Streaming H264 and opus samples

### PR DESCRIPTION
**Pull request by LeoNavel**
This feature adds API for sending H264 and opus samples over RTP and example project.

New API for:

- setting all RTP properties,
- setting all RTCP_SR properties,
- creating RTCP SDES and its chunks,
- RTP packetization,
- RTP packetization of H264 samples (supports single nal unit and nal unit fragmentation),
- RTP packetization of opus samples,
- recording RTP stats and sending RTCP sender report,
- handlers for automatic H264 and opus packetization and for sending RTCP sender reports.

Example project contains:

- source codes for server app to stream video and audio to multiple clients,
- H264 samples (including original video track),
- opus samples (including original audio track),
- python script to generate H264 samples from video,
- python script to generate opus samples from audio,
- html page for client.

### CODE REVIEW

**Comment by paullouisageneau** [comment]
This looks great, and thanks for taking the time to fix the builds! I'm having at look.

File: src/h264packetizationhandler.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "h264packetizationhandler.hpp"
+
+#if RTC_ENABLE_MEDIA
+
+using namespace rtc;
```
Comment (paullouisageneau) [inline src/h264packetizationhandler.cpp:23]: I don't really like `using namespace` to define methods as it kind of defeats the purpose of namespaces (of course it's fine for actually using methods). I prefer defining them in the namespace with `namespace rtc { ... }`.

**Review by paullouisageneau — COMMENTED** [review]
This is awesome work! I've made a few high-level comments, I might make some more during a second pass.

File: examples/streamer/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+else()
+add_executable(streamer main.cpp dispatchqueue.cpp dispatchqueue.hpp h264fileparser.cpp h264fileparser.hpp helpers.cpp helpers.hpp opusfileparser.cpp opusfileparser.hpp fileparser.cpp fileparser.hpp stream.cpp stream.hpp ArgParser.hpp ArgParser.cpp)
+endif()
+set_target_properties(streamer PROPERTIES
+	CXX_STANDARD 17
+	OUTPUT_NAME streamer)
+
+if(WIN32)
+	target_link_libraries(streamer datachannel-static) # DLL exports only the C API
+else()
```
[the remark this answers is not in the document]
Comment (LeoNavel) [reply examples/streamer/CMakeLists.txt]: Fixed

File: include/rtc/h264packetizationhandler.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+#ifndef H264PacketizationHandler_hpp
+#define H264PacketizationHandler_hpp
+
+#include "rtcp.hpp"
+#include "h264rtppacketizer.hpp"
+#include "rtcpsenderreportable.hpp"
+#include "nalunit.hpp"
+
+#if RTC_ENABLE_MEDIA
```
[the remark this answers is not in the document]
Comment (LeoNavel) [reply include/rtc/h264packetizationhandler.hpp]: Fixed

File: src/h264packetizationhandler.cpp (hunk 1 above)
Comment (LeoNavel) [reply src/h264packetizationhandler.cpp]: Fixed

**Comment by LeoNavel** [comment]
@paullouisageneau Do you have any idea how to fix build on windows?

**Comment by paullouisageneau** [comment]
Thank you very much for the changes! I'll have a look soon, I don't have a lot of time these days for personal reasons.

**Comment by paullouisageneau** [comment]
The linking issue on Windows comes from the use of `rtc::make_message` in the new streamer example. It isn't exported for now, so you should declare `RTC_CPP_EXPORT` the non-template `make_message` functions in `include/rtc/message.hpp`. You could also do the same for `to_variant`.

**Comment by stazio** [comment]
Hey @LeoNavel, since you're adding support for SDES packets, may I recommend you process them in `PeerConnection::forwardMedia`?

**Comment by LeoNavel** [comment]
Adding RTC_CPP_EXPORT fixed linking problem. Thanks for that 👍. I added some SDES processing too. Is this PR ready or is there anything else to fix? :)

**Review by paullouisageneau — APPROVED** [review]
It looks good to me, thank you for the work!

**Comment by paullouisageneau** [comment]
@stazio if you don't have remarks on the SDES processing I'm merging this.

**Comment by stazio** [comment]
Looks good to me!

**Comment by stazio** [comment]
Well, I have a slight concern over the potential of someone sending malformed packets with bad lengths. Is there a potential here for a segfault if someone lied about the length of a chunk or item?

**Comment by LeoNavel** [comment]
@stazio You were right. There was potential of segfault. It should be ok now.

File: include/rtc/rtp.hpp [hunk 4]
```diff
@@ -374,8 +412,37 @@ struct RTCP_SDES {
         header.prepareHeader(202, chunkCount, length);
     }
 
+    bool isValid() {
+        auto chunksSize = header.lengthInBytes() - sizeof(header);
+        if (chunksSize == 0) {
+            return true;
+        } else {
+            // there is at least one chunk
+            unsigned int i = 0;
```
Comment (stazio) [inline include/rtc/rtp.hpp:421]: You never increment this value.

Comment (LeoNavel) [reply include/rtc/rtp.hpp:421]: Thanks.

File: include/rtc/rtp.hpp [hunk 5]
```diff
@@ -335,6 +338,39 @@ struct RTCP_SDES_CHUNK {
...
+            auto realSize = RTCP_SDES_CHUNK::size(textsLength);
+            if (realSize > maxChunkSize) {
+                // Chunk is too large
+                return -1;
+            }
+            return realSize;
+        }
+    }
+
     [[nodiscard]] static unsigned int size(std::vector<uint8_t> textLengths) {
```
Comment (stazio) [inline include/rtc/rtp.hpp:374]: This should probably be a const reference to avoid the copy.

Comment (LeoNavel) [reply include/rtc/rtp.hpp]: Fixed.

**Comment by LeoNavel** [comment]
Is there anything else to fix?

**Comment by paullouisageneau** [comment]
@LeoNavel It's all good, could you please resolve the conflict in `peerconnection.cpp` and I'll merge it.

**Comment by LeoNavel** [comment]
Warning about unknown packet is replaced with `COUNTER_UNKNOWN_PACKET_TYPE++` in master. Should I replace warning about invalid SDES packet with `COUNTER_UNKNOWN_PACKET_TYPE++` or add new counter (something like `COUNTER_INVALID_PACKET`) or leave warning as is and just resolve conflict?

**Comment by paullouisageneau** [comment]
> `PLOG_WARNING` about unknown packet is replaced with `COUNTER_UNKNOWN_PACKET_TYPE++` in master. Should I replace warning about invalid SDES packet with `COUNTER_UNKNOWN_PACKET_TYPE++` or add new counter (something like `COUNTER_INVALID_PACKET`) or leave warning as is and just resolve conflict?

Oh, good point, a `COUNTER_INVALID_PACKET` would be perfect! Otherwise just leave the warning and I'll add the counter later.

**Comment by LeoNavel** [comment]
@paullouisageneau Do you know why test with libnice / build-media failed?

**Comment by paullouisageneau** [comment]
> @paullouisageneau Do you know why test with libnice / build-media failed?

This specific test sometimes fails one the CI, I have to investigate it. I relaunched it and it passed fine. I'm merging, thanks again for the work!

**Pull request merged**